pkg/reconcilermanager/controllers/secret.go

```diff
@@ -60,69 +60,86 @@ func shouldUpsertHelmSecret(rs *v1beta1.RepoSync) bool {
 
 // upsertSecrets creates or updates all secrets in config-management-system
 // namespace using existing secrets in the reposync.namespace.
-func upsertSecrets(ctx context.Context, rs *v1beta1.RepoSync, c client.Client, reconcilerName string) error {
+func upsertSecrets(ctx context.Context, rs *v1beta1.RepoSync, c client.Client, reconcilerKey types.NamespacedName) (client.ObjectKey, error) {
 	if shouldUpsertPrivateCertSecret(rs) {
-		err := upsertSecret(ctx, rs, c, reconcilerName, rs.Spec.Git.PrivateCertSecret.Name)
+		sKey, err := upsertSecret(ctx, rs, c, reconcilerKey, rs.Spec.Git.PrivateCertSecret.Name)
 		if err != nil {
-			return err
+			return sKey, err
 		}
+		return sKey, nil
 	}
 	if shouldUpsertGitSecret(rs) {
-		err := upsertSecret(ctx, rs, c, reconcilerName, rs.Spec.Git.SecretRef.Name)
+		sKey, err := upsertSecret(ctx, rs, c, reconcilerKey, rs.Spec.Git.SecretRef.Name)
 		if err != nil {
-			return err
+			return sKey, err
 		}
+		return sKey, nil
 	}
 	if shouldUpsertHelmSecret(rs) {
-		err := upsertSecret(ctx, rs, c, reconcilerName, rs.Spec.Helm.SecretRef.Name)
+		sKey, err := upsertSecret(ctx, rs, c, reconcilerKey, rs.Spec.Helm.SecretRef.Name)
 		if err != nil {
-			return err
+			return sKey, err
 		}
+		return sKey, nil
 	}
-	return nil
+	// No secret required
+	return client.ObjectKey{}, nil
 }
 
 // upsertSecret creates or updates a secret in config-management-system
 // namespace using an existing secret in the reposync.namespace.
-func upsertSecret(ctx context.Context, rs *v1beta1.RepoSync, c client.Client, reconcilerName, nsSecretName string) error {
-	// namespaceSecret represent secret in reposync.namespace.
-	namespaceSecret := &corev1.Secret{}
-	if err := get(ctx, nsSecretName, rs.Namespace, namespaceSecret, c); err != nil {
+func upsertSecret(ctx context.Context, rs *v1beta1.RepoSync, c client.Client, reconcilerKey types.NamespacedName, nsSecretName string) (client.ObjectKey, error) {
+	// cmsSecretKey represents the secret in config-management-system namespace,
+	// managed by the reconciler-manager.
+	// This is the object being upserted, and should be returned.
+	cmsSecretKey := client.ObjectKey{
+		Namespace: reconcilerKey.Namespace,
+		Name:      ReconcilerResourceName(reconcilerKey.Name, nsSecretName),
+	}
+	// nsSecretKey represents the secret in the RepoSync namespace,
+	// managed by the user.
+	nsSecretKey := client.ObjectKey{
+		Namespace: rs.Namespace,
+		Name:      nsSecretName,
+	}
+	nsSecret := &corev1.Secret{}
+	if err := getSecret(ctx, c, nsSecretKey, nsSecret); err != nil {
 		if apierrors.IsNotFound(err) {
-			return errors.Errorf(
-				"%s not found. Create %s secret in %s namespace", nsSecretName, nsSecretName, rs.Namespace)
+			return cmsSecretKey, errors.Errorf(
+				"secret %s not found: create one to allow client authentication", nsSecretKey)
```

Comment:
This may be too specific in the case of the certificate secret, as the secret is for certificate verification rather than authentication

Reply:
AFAICT, `.spec.git.privateCertSecret` is a private key for git authentication, not a public key for certificate verification.

Tho the API for it is pretty terrible, IMO. I would have used the same  `.spec.git.secretRef` and a `.spec.git.secretType` or something to indicate it was a cert and not un/pw.

Reply:
I've split upsertSecrets into upsertAuthSecret and upsertCACertSecret to make it easier to wrap the errors with more actionable error messages, and make it clear that the auth secrets and mutually exclusive, but the ca cert is extra and may co-exist with the git secret.